
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../api/misc/">
      
      
        <link rel="next" href="../cite_us/">
      
      
      <link rel="icon" href="../assets/sopa_favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.14">
    
    
      
        <title>Frequently asked questions - Sopa</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.342714a4.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../assets/_mkdocstrings.css">
    
      <link rel="stylesheet" href="../css/termynal.css">
    
      <link rel="stylesheet" href="../css/custom.css">
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#what-are-the-inputs-of-sopa" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

  

<header class="md-header md-header--shadow" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="Sopa" class="md-header__button md-logo" aria-label="Sopa" data-md-component="logo">
      
  <img src="../assets/sopa_small_white.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Sopa
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Frequently asked questions
            
          </span>
        </div>
      </div>
    </div>
    
      
    
    
    
    
      
      
        <label class="md-header__button md-icon" for="__search">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        </label>
        <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
      
    
    
      <div class="md-header__source">
        <a href="https://github.com/gustaveroussy/sopa" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    gustaveroussy/sopa
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    



<nav class="md-nav md-nav--primary" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="Sopa" class="md-nav__button md-logo" aria-label="Sopa" data-md-component="logo">
      
  <img src="../assets/sopa_small_white.png" alt="logo">

    </a>
    Sopa
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/gustaveroussy/sopa" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    gustaveroussy/sopa
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../getting_started/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Getting Started
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" >
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Tutorials
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Tutorials
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/api_usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    API usage
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/snakemake/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Snakemake pipeline
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/techno_specific/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Tech-specific advice
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/xenium_explorer/explorer/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Xenium Explorer usage
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/align/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Landmark-based alignment
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/compare_segmentations/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Comparing segmentations
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/spatial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Spatial operations
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/visium_hd/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Visium HD tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/he/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    H&amp;E usage
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/cli_usage/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Command line interface
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3_11" >
        
          
          <label class="md-nav__link" for="__nav_3_11" id="__nav_3_11_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    Other segmentations
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_11_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_11">
            <span class="md-nav__icon md-icon"></span>
            Other segmentations
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/comseg/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    ComSeg usage
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/multi_step_segmentation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Multi-step
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../tutorials/custom_segmentation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Custom segmentation
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
    
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  
  <span class="md-ellipsis">
    API
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            API
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/readers/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Readers
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/patches/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Patches
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/segmentation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Segmentation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/aggregation/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Aggregation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/spatial/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Spatial operations
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/utils/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Utils
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../api/misc/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Misc
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  
  <span class="md-ellipsis">
    Frequently asked questions
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  
  <span class="md-ellipsis">
    Frequently asked questions
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#what-are-the-inputs-of-sopa" class="md-nav__link">
    <span class="md-ellipsis">
      What are the inputs of Sopa?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-disable-the-auto-save" class="md-nav__link">
    <span class="md-ellipsis">
      How to disable the auto-save?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-parallelize-segmentation" class="md-nav__link">
    <span class="md-ellipsis">
      How to parallelize segmentation?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#which-pipeline-parameters-should-i-use" class="md-nav__link">
    <span class="md-ellipsis">
      Which pipeline parameters should I use?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-filter-genes" class="md-nav__link">
    <span class="md-ellipsis">
      How to filter genes?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-does-sopa-know-when-using-which-elements" class="md-nav__link">
    <span class="md-ellipsis">
      How does Sopa know when using which elements?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-remove-cells-artifacts" class="md-nav__link">
    <span class="md-ellipsis">
      How to remove cells artifacts?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-get-more-cells-with-cellpose" class="md-nav__link">
    <span class="md-ellipsis">
      How to get more cells with Cellpose?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-use-a-custom-cellpose-model" class="md-nav__link">
    <span class="md-ellipsis">
      How to use a custom Cellpose model?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-use-the-gpu-for-cellpose" class="md-nav__link">
    <span class="md-ellipsis">
      How to use the GPU for Cellpose?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-provide-other-arguments-to-cellpose" class="md-nav__link">
    <span class="md-ellipsis">
      How to provide other arguments to Cellpose?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-use-a-prior-cell-segmentation" class="md-nav__link">
    <span class="md-ellipsis">
      How to use a prior cell segmentation?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-optimize-the-segmentation-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      How to optimize the segmentation parameters?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-provide-dictionnaries-to-cli-arguments" class="md-nav__link">
    <span class="md-ellipsis">
      How to provide dictionnaries to CLI arguments?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-fix-an-out-of-memory-issue-on-merscope-data" class="md-nav__link">
    <span class="md-ellipsis">
      How to fix an "out-of-memory" issue on MERSCOPE data?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-remove-the-logs" class="md-nav__link">
    <span class="md-ellipsis">
      How to remove the logs?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-ask-for-help" class="md-nav__link">
    <span class="md-ellipsis">
      How to ask for help?
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../cite_us/" class="md-nav__link">
        
  
  
  <span class="md-ellipsis">
    Cite us
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#what-are-the-inputs-of-sopa" class="md-nav__link">
    <span class="md-ellipsis">
      What are the inputs of Sopa?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-disable-the-auto-save" class="md-nav__link">
    <span class="md-ellipsis">
      How to disable the auto-save?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-parallelize-segmentation" class="md-nav__link">
    <span class="md-ellipsis">
      How to parallelize segmentation?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#which-pipeline-parameters-should-i-use" class="md-nav__link">
    <span class="md-ellipsis">
      Which pipeline parameters should I use?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-filter-genes" class="md-nav__link">
    <span class="md-ellipsis">
      How to filter genes?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-does-sopa-know-when-using-which-elements" class="md-nav__link">
    <span class="md-ellipsis">
      How does Sopa know when using which elements?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-remove-cells-artifacts" class="md-nav__link">
    <span class="md-ellipsis">
      How to remove cells artifacts?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-get-more-cells-with-cellpose" class="md-nav__link">
    <span class="md-ellipsis">
      How to get more cells with Cellpose?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-use-a-custom-cellpose-model" class="md-nav__link">
    <span class="md-ellipsis">
      How to use a custom Cellpose model?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-use-the-gpu-for-cellpose" class="md-nav__link">
    <span class="md-ellipsis">
      How to use the GPU for Cellpose?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-provide-other-arguments-to-cellpose" class="md-nav__link">
    <span class="md-ellipsis">
      How to provide other arguments to Cellpose?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-use-a-prior-cell-segmentation" class="md-nav__link">
    <span class="md-ellipsis">
      How to use a prior cell segmentation?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-optimize-the-segmentation-parameters" class="md-nav__link">
    <span class="md-ellipsis">
      How to optimize the segmentation parameters?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-provide-dictionnaries-to-cli-arguments" class="md-nav__link">
    <span class="md-ellipsis">
      How to provide dictionnaries to CLI arguments?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-fix-an-out-of-memory-issue-on-merscope-data" class="md-nav__link">
    <span class="md-ellipsis">
      How to fix an "out-of-memory" issue on MERSCOPE data?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-remove-the-logs" class="md-nav__link">
    <span class="md-ellipsis">
      How to remove the logs?
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#how-to-ask-for-help" class="md-nav__link">
    <span class="md-ellipsis">
      How to ask for help?
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


  <h1>Frequently asked questions</h1>

<h2 id="what-are-the-inputs-of-sopa">What are the inputs of Sopa?</h2>
<p>You need the raw inputs of your machine, that is:</p>
<ul>
<li>
<p>One or multiple image(s), usually corresponding to one or multiple <code>.tiff</code> file(s)</p>
</li>
<li>
<p>Optionally, a file of transcript location, usually a <code>.csv</code> or <code>.parquet</code> file</p>
</li>
</ul>
<p>In this documentation, <code>data_path</code> denotes the path to your raw data. Select the correct tab below to understand what is the right path to your raw data:</p>
<div class="tabbed-set tabbed-alternate" data-tabs="1:7"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><input id="__tabbed_1_2" name="__tabbed_1" type="radio" /><input id="__tabbed_1_3" name="__tabbed_1" type="radio" /><input id="__tabbed_1_4" name="__tabbed_1" type="radio" /><input id="__tabbed_1_5" name="__tabbed_1" type="radio" /><input id="__tabbed_1_6" name="__tabbed_1" type="radio" /><input id="__tabbed_1_7" name="__tabbed_1" type="radio" /><div class="tabbed-labels"><label for="__tabbed_1_1">Xenium</label><label for="__tabbed_1_2">MERSCOPE</label><label for="__tabbed_1_3">CosMx</label><label for="__tabbed_1_4">MACSima</label><label for="__tabbed_1_5">PhenoCycler</label><label for="__tabbed_1_6">Hyperion</label><label for="__tabbed_1_7">Others (CZI, ...)</label></div>
<div class="tabbed-content">
<div class="tabbed-block">
<p><code>data_path</code> is the path to the directory containing the following files: <code>morphology.ome.tif</code>, <code>experiment.xenium</code> and <code>transcripts.parquet</code>. In brief, you should have this file structure:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a>.
<a id="__codelineno-0-2" name="__codelineno-0-2" href="#__codelineno-0-2"></a>├─ morphology_focus.ome.tif   # or a directory (for recent versions of the Xenium)
<a id="__codelineno-0-3" name="__codelineno-0-3" href="#__codelineno-0-3"></a>├─ experiment.xenium
<a id="__codelineno-0-4" name="__codelineno-0-4" href="#__codelineno-0-4"></a>└─ transcripts.parquet
</code></pre></div>
</div>
<div class="tabbed-block">
<p><code>data_path</code> is the path to the "region" directory containing a <code>detected_transcripts.csv</code> file and an <code>images</code> directory. For instance, the directory may be called <code>region_0</code>. In brief, you should have this file structure:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a>.
<a id="__codelineno-1-2" name="__codelineno-1-2" href="#__codelineno-1-2"></a>├─ detected_transcripts.csv
<a id="__codelineno-1-3" name="__codelineno-1-3" href="#__codelineno-1-3"></a>└─ images
<a id="__codelineno-1-4" name="__codelineno-1-4" href="#__codelineno-1-4"></a>   ├─ mosaic_{stain}_z{z_layer}.tif
<a id="__codelineno-1-5" name="__codelineno-1-5" href="#__codelineno-1-5"></a>   └─ micron_to_mosaic_pixel_transform.csv
</code></pre></div>
</div>
<div class="tabbed-block">
<p><code>data_path</code> is the path to the directory containing:</p>
<ul>
<li>a transcript file <code>*_tx_file</code> (with columns <code>target</code>, <code>x_global_px</code>, <code>y_global_px</code>)</li>
<li>a FOV locations file <code>*_fov_positions_file</code> (with columns <code>FOV</code>, <code>X_mm</code>, <code>Y_mm</code>)</li>
<li>a <code>Morphology_ChannelID_Dictionary.txt</code> file containing channel names</li>
<li>a <code>Morphology2D</code> directory containing the images, end in <code>_F*.TIF</code>.</li>
</ul>
<p>These files must be exported as flat files in AtomX. That is: within a study, click on "Export" and then select files from the "Flat CSV Files" section (transcripts flat and FOV position flat). You should have this file structure:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a>.
<a id="__codelineno-2-2" name="__codelineno-2-2" href="#__codelineno-2-2"></a>├─ &lt;DATASET_ID&gt;_tx_file.csv (or csv.gz)
<a id="__codelineno-2-3" name="__codelineno-2-3" href="#__codelineno-2-3"></a>├─ &lt;DATASET_ID&gt;_fov_positions_file.csv (or csv.gz)
<a id="__codelineno-2-4" name="__codelineno-2-4" href="#__codelineno-2-4"></a>├─ Morphology_ChannelID_Dictionary.txt
<a id="__codelineno-2-5" name="__codelineno-2-5" href="#__codelineno-2-5"></a>└─ Morphology2D
<a id="__codelineno-2-6" name="__codelineno-2-6" href="#__codelineno-2-6"></a>   ├─ XXX_F001.TIF
<a id="__codelineno-2-7" name="__codelineno-2-7" href="#__codelineno-2-7"></a>   ├─ XXX_F002.TIF
<a id="__codelineno-2-8" name="__codelineno-2-8" href="#__codelineno-2-8"></a>   └─ ...
</code></pre></div>
</div>
<div class="tabbed-block">
<p><code>data_path</code> is the path to the directory containing multiple <code>.ome.tif</code> files (one file per channel). In brief, you should have this file structure:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a>.
<a id="__codelineno-3-2" name="__codelineno-3-2" href="#__codelineno-3-2"></a>├─ AAA.ome.tif
<a id="__codelineno-3-3" name="__codelineno-3-3" href="#__codelineno-3-3"></a>├─ BBB.ome.tif
<a id="__codelineno-3-4" name="__codelineno-3-4" href="#__codelineno-3-4"></a>└─ CCC.ome.tif
</code></pre></div>
</div>
<div class="tabbed-block">
<p><code>data_path</code> is the path to one <code>.qptiff</code> file, or one <code>.tif</code> file (if exported from QuPath).</p>
</div>
<div class="tabbed-block">
<p><code>data_path</code> is path to the directory containing multiple <code>.ome.tiff</code> files (one file per channel). In brief, you should have this file structure:
<div class="highlight"><pre><span></span><code><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a>.
<a id="__codelineno-4-2" name="__codelineno-4-2" href="#__codelineno-4-2"></a>├─ AAA.ome.tiff
<a id="__codelineno-4-3" name="__codelineno-4-3" href="#__codelineno-4-3"></a>├─ BBB.ome.tiff
<a id="__codelineno-4-4" name="__codelineno-4-4" href="#__codelineno-4-4"></a>└─ CCC.ome.tiff
</code></pre></div></p>
</div>
<div class="tabbed-block">
<p>Other file formats (ND2, CZI, LIF, or DV) are supported via the <code>bioio</code> reader. In that case, you'll need to add new dependencies: <code>pip install bioio</code> (and potentially some file-format specific dependencies, see <a href="https://bioio-devs.github.io/bioio/OVERVIEW.html#reader-installation">documentation</a>).</p>
<p>This reader is called <code>bioio</code>, i.e. you can use it via <code>sopa.io.bioio(data_path)</code>, where <code>data_path</code> is the path to your data file containing your image(s). For the Snakemake pipeline, provide <code>bioio</code> as a <code>technology</code> in the config file.</p>
</div>
</div>
</div>
<h2 id="how-to-disable-the-auto-save">How to disable the auto-save?</h2>
<p>When using the API, and when your <code>SpatialData</code> object is saved on-disk, Sopa will automatically save any new spatial element on disk by default. To disable this behavior, use the following command from the API:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="n">sopa</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">auto_save_on_disk</span> <span class="o">=</span> <span class="kc">False</span>
</code></pre></div>
<h2 id="how-to-parallelize-segmentation">How to parallelize segmentation?</h2>
<p>Some steps of Sopa, notably the segmentation, can be accelerated via a parallelization backend. If you use the API, you can set the <code>"dask"</code> backend as below.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="c1"># when using the API</span>
<a id="__codelineno-6-2" name="__codelineno-6-2" href="#__codelineno-6-2"></a><span class="n">sopa</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">parallelization_backend</span> <span class="o">=</span> <span class="s2">&quot;dask&quot;</span>
</code></pre></div>
<p>Otherwise, if you don't use the API, you can also set the <code>SOPA_PARALLELIZATION_BACKEND</code> env variable, e.g.:
<div class="highlight"><pre><span></span><code><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="nb">export</span><span class="w"> </span><span class="nv">SOPA_PARALLELIZATION_BACKEND</span><span class="o">=</span>dask
</code></pre></div></p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The <code>dask</code> backend is still experimental. You can add a comment to <a href="https://github.com/gustaveroussy/sopa/issues/145">this issue</a> to help us improve it.</p>
</div>
<p>You can also pass some kwargs to the <a href="https://distributed.dask.org/en/stable/api.html#client">dask Client</a> as below. These kwargs are highly dependent of your cluster and the size of your patches. For "middle-size" patches, we recommend about 4GB of memory per worker for cellpose, and between 8GB and 16GB or memory for baysor.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a><span class="n">sopa</span><span class="o">.</span><span class="n">settings</span><span class="o">.</span><span class="n">dask_client_kwargs</span><span class="p">[</span><span class="s2">&quot;n_workers&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">4</span>
</code></pre></div>
<p>For testing purposes, you can run the lines below, which will show you how many workers and memory you have by default:
<div class="highlight"><pre><span></span><code><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="kn">from</span><span class="w"> </span><span class="nn">dask.distributed</span><span class="w"> </span><span class="kn">import</span> <span class="n">Client</span>
<a id="__codelineno-9-2" name="__codelineno-9-2" href="#__codelineno-9-2"></a>
<a id="__codelineno-9-3" name="__codelineno-9-3" href="#__codelineno-9-3"></a><span class="n">client</span> <span class="o">=</span> <span class="n">Client</span><span class="p">()</span>
<a id="__codelineno-9-4" name="__codelineno-9-4" href="#__codelineno-9-4"></a>
<a id="__codelineno-9-5" name="__codelineno-9-5" href="#__codelineno-9-5"></a><span class="n">n_workers</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">client</span><span class="o">.</span><span class="n">cluster</span><span class="o">.</span><span class="n">workers</span><span class="p">)</span>
<a id="__codelineno-9-6" name="__codelineno-9-6" href="#__codelineno-9-6"></a><span class="n">mem_worker0</span> <span class="o">=</span> <span class="n">client</span><span class="o">.</span><span class="n">cluster</span><span class="o">.</span><span class="n">workers</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">memory_manager</span><span class="o">.</span><span class="n">memory_limit</span> <span class="o">/</span> <span class="mi">1024</span><span class="o">**</span><span class="mi">3</span>
<a id="__codelineno-9-7" name="__codelineno-9-7" href="#__codelineno-9-7"></a>
<a id="__codelineno-9-8" name="__codelineno-9-8" href="#__codelineno-9-8"></a><span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">n_workers</span><span class="si">=}</span><span class="s2">, </span><span class="si">{</span><span class="n">mem_worker0</span><span class="si">=:</span><span class="s2">.3</span><span class="si">}</span><span class="s2">GB&quot;</span><span class="p">)</span>
</code></pre></div></p>
<h2 id="which-pipeline-parameters-should-i-use">Which pipeline parameters should I use?</h2>
<p>Some parameters such as the Cellpose diameter is crucial and depends highly on the resolution of your technology (pixel size).
As a guide, you can start from the parameters of the <a href="https://github.com/gustaveroussy/sopa/tree/main/workflow/config">config files</a> of your specific technology, and adjust them based on your knowledge of the tissue you work on.</p>
<p>Here are some parameters which are important to check: <code>diameter</code> (in pixels for cellpose), <code>min_area</code> (in pixels^2 for cellpose, in microns^2 for baysor), <code>scale</code> (in microns for baysor), <code>pixel_size</code> (for the Xenium Explorer conversion, to have the right scale during display).</p>
<h2 id="how-to-filter-genes">How to filter genes?</h2>
<p>By default, we remove some genes names during segmentation and aggregation (for instance, <code>"blank"</code> or <code>"unassigned"</code> gene names). To change this behavior, you can update the gene pattern under <code>sopa.settings.gene_exclude_pattern</code> (this pattern is used by <a href="https://pandas.pydata.org/docs/reference/api/pandas.Series.str.match.html"><code>pandas.Series.str.match</code></a>).</p>
<p>You can also decide not to remove some specific low quality transcripts for segmenation. To do that, create a (boolean) column called <code>"low_quality_transcript"</code> to your transcript dataframe. The rows whose value is <code>True</code> will not be used during segmentation. For instance, if you have Xenium data, you can filter the genes based on a QV value of 20.</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="n">df</span> <span class="o">=</span> <span class="n">sdata</span><span class="p">[</span><span class="s2">&quot;transcripts&quot;</span><span class="p">]</span>
<a id="__codelineno-10-2" name="__codelineno-10-2" href="#__codelineno-10-2"></a><span class="n">df</span><span class="p">[</span><span class="s2">&quot;low_quality_transcript&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">qv</span> <span class="o">&lt;</span> <span class="mi">20</span>
</code></pre></div>
<h2 id="how-does-sopa-know-when-using-which-elements">How does Sopa know when using which elements?</h2>
<p>Many functions of Sopa can run without any argument. For instance, <code>sopa.aggregate(sdata)</code> works for very different technologies, such as VisiumHD, MERSCOPE, or MACSima data.</p>
<p>Internally, when reading the raw data, Sopa saves some attributes inside <code>sdata.attrs</code>. These attributes are then used to know which spatial element corresponds to what. For instance, one H&amp;E image may be tagged for tissue segmentation, while the DAPI image will be tagged to be used for cellpose.</p>
<p>Sopa handles this internally by default, and it is completely invisible to the user. But, to get a full control on what is done, you can always set some arguments to specify which element to be used. You can refer to the following arguments of the API: <code>image_key</code>, <code>points_key</code>, <code>shapes_key</code>, <code>bins_key</code>, and <code>table_key</code>.</p>
<h2 id="how-to-remove-cells-artifacts">How to remove cells artifacts?</h2>
<p>When segmenting a patch that is outside of the issue, Cellpose may "hallucinate" and generate some fake cells. To avoid that, you can run <a href="../api/segmentation/#sopa.segmentation.tissue"><code>sopa.segmentation.tissue</code></a> to ensure segmentation is always run inside the tissue.</p>
<p>Otherwise, if you have inside the tissue some small cells artefacts, <code>Sopa</code> offers three filtering approaches:</p>
<ul>
<li>Using a <a href="../api/segmentation/#sopa.segmentation.cellpose">min_area</a> threshold to remove small cells (provide this argument to the segmentation methods).</li>
<li>A <a href="/api/aggregation/#sopa.aggregate">min_transcripts</a> threshold to remove cells with a low transcript count (provide this argument to the aggregation step).</li>
<li>A <a href="..//api/aggregation/#sopa.aggregate">min_intensity_ratio</a> value to remove cells with a low fluorescence intensity (provide this argument to the aggregation step).</li>
</ul>
<h2 id="how-to-get-more-cells-with-cellpose">How to get more cells with Cellpose?</h2>
<ul>
<li>The main Cellpose parameter to check is <code>diameter</code>, i.e. a typical cell diameter <strong>in pixels</strong>. Note that this is highly specific to the technology you're using since the micron-to-pixel ratio can differ. We advise you to start with the default parameter for your technology of interest (see the <code>diameter</code> parameter inside our config files <a href="https://github.com/gustaveroussy/sopa/tree/main/workflow/config">here</a>).</li>
<li>Maybe <code>min_area</code> is too high, and all the cells are filtered because they are smaller than this area. Remind that, when using Cellpose, the areas correspond to pixels^2.</li>
<li>This can be due to a low image quality. If the image is too pixelated, consider increasing <code>gaussian_sigma</code> (e.g., <code>2</code>) under the cellpose parameters of our config. If the image has a low contrast, consider increasing <code>clip_limit</code> (e.g., <code>0.3</code>). These parameters are detailed in <a href="https://github.com/gustaveroussy/sopa/blob/main/workflow/config/example_commented.yaml">this example config</a>.</li>
<li>Consider updating the official Cellpose parameters. In particular, try <code>cellprob_threshold=-6</code> and <code>flow_threshold=2</code>.</li>
</ul>
<h2 id="how-to-use-a-custom-cellpose-model">How to use a custom Cellpose model?</h2>
<p>You can use any existing <a href="https://cellpose.readthedocs.io/en/latest/models.html">Cellpose model</a> with the <code>model_type</code> argument (via the API, CLI, or Snakemake pipeline). For the Snakemake pipeline, see <a href="https://github.com/gustaveroussy/sopa/blob/main/workflow/config/example_commented.yaml">here</a> how to set this argument.
If you have a custom pretrained model, use the <code>pretrained_model</code> argument instead of <code>model_type</code>, and give the path to your cellpose model.</p>
<h2 id="how-to-use-the-gpu-for-cellpose">How to use the GPU for Cellpose?</h2>
<p>You can provide <code>gpu=True</code> to <code>sopa.segmentation.cellpose</code>. This is recommended for <code>cellpose&gt;=4.0.0</code>, which supports larger models and may be much faster on GPUs.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>If you have many CPU cores and only one GPU, it may be faster to run in parallel on CPUs rather than sequentially using the GPU (mostly on <code>cellpose&lt;4.0.0</code>). Also, if you are on MacOS, you may experience issues because the PyTorch MPS backend doesn't support all features yet.</p>
</div>
<div class="highlight"><pre><span></span><code><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">sopa</span>
<a id="__codelineno-11-2" name="__codelineno-11-2" href="#__codelineno-11-2"></a>
<a id="__codelineno-11-3" name="__codelineno-11-3" href="#__codelineno-11-3"></a><span class="n">sdata</span> <span class="o">=</span> <span class="n">sopa</span><span class="o">.</span><span class="n">io</span><span class="o">.</span><span class="n">toy_dataset</span><span class="p">()</span>
<a id="__codelineno-11-4" name="__codelineno-11-4" href="#__codelineno-11-4"></a><span class="n">sopa</span><span class="o">.</span><span class="n">make_image_patches</span><span class="p">(</span><span class="n">sdata</span><span class="p">)</span>
<a id="__codelineno-11-5" name="__codelineno-11-5" href="#__codelineno-11-5"></a>
<a id="__codelineno-11-6" name="__codelineno-11-6" href="#__codelineno-11-6"></a><span class="n">sopa</span><span class="o">.</span><span class="n">segmentation</span><span class="o">.</span><span class="n">cellpose</span><span class="p">(</span><span class="n">sdata</span><span class="p">,</span> <span class="n">channels</span><span class="o">=</span><span class="s2">&quot;DAPI&quot;</span><span class="p">,</span> <span class="n">diameter</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">cellpose_model_kwargs</span><span class="o">=</span><span class="p">{</span><span class="s2">&quot;gpu&quot;</span><span class="p">:</span> <span class="kc">True</span><span class="p">})</span>
</code></pre></div>
<h2 id="how-to-provide-other-arguments-to-cellpose">How to provide other arguments to Cellpose?</h2>
<p>When using the Snakemake pipeline, you can use <code>method_kwargs</code> to provide extra arguments to Cellpose. For instance, we use <code>resample=False</code> in the example below, which may significantly speed up the segmentation while not decreasing significantly the segmentation quality:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a><span class="nt">segmentation</span><span class="p">:</span>
<a id="__codelineno-12-2" name="__codelineno-12-2" href="#__codelineno-12-2"></a><span class="w">  </span><span class="nt">cellpose</span><span class="p">:</span>
<a id="__codelineno-12-3" name="__codelineno-12-3" href="#__codelineno-12-3"></a><span class="w">    </span><span class="nt">diameter</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">60</span>
<a id="__codelineno-12-4" name="__codelineno-12-4" href="#__codelineno-12-4"></a><span class="w">    </span><span class="nt">channels</span><span class="p">:</span><span class="w"> </span><span class="p p-Indicator">[</span><span class="s">&quot;DAPI&quot;</span><span class="p p-Indicator">]</span>
<a id="__codelineno-12-5" name="__codelineno-12-5" href="#__codelineno-12-5"></a><span class="w">    </span><span class="nt">flow_threshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2</span>
<a id="__codelineno-12-6" name="__codelineno-12-6" href="#__codelineno-12-6"></a><span class="w">    </span><span class="nt">cellprob_threshold</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">-6</span>
<a id="__codelineno-12-7" name="__codelineno-12-7" href="#__codelineno-12-7"></a><span class="w">    </span><span class="nt">min_area</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">2000</span>
<a id="__codelineno-12-8" name="__codelineno-12-8" href="#__codelineno-12-8"></a><span class="w">    </span><span class="nt">method_kwargs</span><span class="p">:</span>
<a id="__codelineno-12-9" name="__codelineno-12-9" href="#__codelineno-12-9"></a><span class="w">      </span><span class="nt">resample</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">False</span>
</code></pre></div>
<h2 id="how-to-use-a-prior-cell-segmentation">How to use a prior cell segmentation?</h2>
<p>If you have MERSCOPE or Xenium data, you probably already have a cell segmentation. This can be used as a prior for Proseg or Baysor, instead of running Cellpose with Sopa. For that, you have an existing config file for the Snakemake pipeline for both <a href="https://github.com/gustaveroussy/sopa/blob/main/workflow/config/merscope/baysor_vizgen.yaml">MERSCOPE</a> and <a href="https://github.com/gustaveroussy/sopa/blob/main/workflow/config/xenium/baysor_prior.yaml">Xenium</a> data. If using the API/CLI, consider using the <code>prior_shapes_key</code> and the <code>unassigned_value</code> arguments when creating the patches for the transcripts. For MERSCOPE data, <code>prior_shapes_key="cell_id"</code> and <code>unassigned_value=-1</code>. For Xenium data, <code>prior_shapes_key="cell_id"</code> and <code>unassigned_value="UNASSIGNED"</code>. You can also decide to run Cellpose via Sopa, and then use it as a prior: in that case, simply pass <code>prior_shapes_key="cellpose_boundaries"</code> after running cellpose.</p>
<h2 id="how-to-optimize-the-segmentation-parameters">How to optimize the segmentation parameters?</h2>
<p>Selecting the right parameters for Cellpose/Baysor can significantly improve the output data quality. To choose these parameters, we recommend subsetting the data (<code>spatialdata.bounding_box_query</code>), saving the subset (<code>sdata.write("subset.zarr")</code>), running different segmentation on the subset (use <code>key_added</code> to save the segmentation with a specific name), and compare the results. Refer to <a href="../tutorials/compare_segmentations">this tutorial</a> for an example.</p>
<h2 id="how-to-provide-dictionnaries-to-cli-arguments">How to provide dictionnaries to CLI arguments?</h2>
<p>Some CLI arguments are optionnal dictionnaries. For instance, <code>sopa convert</code> has a <code>--kwargs</code> option. In that case, a dictionnary can be provided as an inline string, for instance:</p>
<p><code>--kwargs "{'backend': 'rioxarray'}"</code></p>
<h2 id="how-to-fix-an-out-of-memory-issue-on-merscope-data">How to fix an "out-of-memory" issue on MERSCOPE data?</h2>
<p>If using MERSCOPE data, images can be huge. To improve RAM efficiency, you can install <code>rioxarray</code> (<code>pip install rioxarray</code>). Then, the <code>rioxarray</code> will be used by default by the reader (no change needed, it will be detected automatically).</p>
<h2 id="how-to-remove-the-logs">How to remove the logs?</h2>
<p>You can change the level of <code>logging</code> for sopa, e.g. you can run the lines below to set the logging level to show only errors:</p>
<div class="highlight"><pre><span></span><code><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">sopa</span>
<a id="__codelineno-13-2" name="__codelineno-13-2" href="#__codelineno-13-2"></a>
<a id="__codelineno-13-3" name="__codelineno-13-3" href="#__codelineno-13-3"></a><span class="n">sopa</span><span class="o">.</span><span class="n">log</span><span class="o">.</span><span class="n">setLevel</span><span class="p">(</span><span class="n">sopa</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">ERROR</span><span class="p">)</span>
</code></pre></div>
<h2 id="how-to-ask-for-help">How to ask for help?</h2>
<p>If you have an issue that is not detailed in this FAQ, you can still open an issue on <a href="https://github.com/gustaveroussy/sopa/issues">Sopa's Github repository</a>, and detail your issue with as much precision as possible for the maintainers to be able to reproduce it.</p>
<p>Make sure to have a quick look to the existing issues, maybe someone faced the same problem.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2023 - 2024 Quentin Blampey
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    
      
      <script id="__config" type="application/json">{"base": "..", "features": ["content.code.annotate", "content.code.copy"], "search": "../assets/javascripts/workers/search.d50fe291.min.js", "tags": null, "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": null}</script>
    
    
      <script src="../assets/javascripts/bundle.13a4f30d.min.js"></script>
      
        <script src="../js/termynal.js"></script>
      
        <script src="../js/custom.js"></script>
      
        <script src="../js/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>